### DISCUSSION TITLE
Using Dasel to export JSON at different levels to CSV

### DISCUSSION THREAD

**Discussion by eroth** [TomWright/dasel#273] [Q&A] [answered]
Hi, thanks so much for making Dasel! I'm trying to use it to export JSON at different levels (heterogenous levels?) to CSV and am wondering if it can do what I'm looking for. I've looked a bunch here and in various blog posts but can't seem to find anything or figure it out. This is a sample of the structure I'm working with (an array of these objects pulled from GH's GraphQL API):

``` json
[
{
  "author" : {
    "login" : "querty"
  },
  "number" : 4521,
  "createdAt" : "2022-11-21T02:57:35Z",
  "mergedAt" : "2022-12-05T02:16:18Z",
  "mergedBy" : {
    "login" : "querty"
  },
  "approvers" : {
    "nodes" : [
      {
        "author" : {
          "name" : "Bill Blass",
          "login" : "bblass"
        },
        "state" : "APPROVED"
      },
      {
        "author" : {
          "name" : "Ricky Eckhart",
          "login" : "wreck-hart"
        },
        "state" : "APPROVED"
      }
    ]
  },
  "title" : "Does something really amazing and magical",
  "repository" : {
    "owner" : {
      "login" : "orgo"
    },
    "name" : "my-my-my"
  }
},
{
...
}
]
```

What I'm trying to do is only pull out certain fields at different levels, then export them to CSV. What I'd like to end up with is something like the below, where `author` comes from `author.login` and everything else is at the same level of the structure:
```csv
author,number,title,createdAt,mergedAt
"querty",4521,"Does something really amazing and magical","2022-11-21T02:57:35Z","2022-12-05T02:16:18Z"
...
```

I'm able to print out one value, e.g., all values for `author`, but can't seem to figure out how to map a custom structure out of this and export it to CSV like what I'm after. Any help would be appreciated!

**Comment by TomWright** [comment, marked as answer]
This sounds like a good use-case for [mapOf](https://daseldocs.tomwright.me/functions/mapof).

```
$ dasel -f file.json -w csv 'all().mapOf(author,author.login,number,number,title,title).merge()'
author,number,title
querty,4521.000000,Does something really amazing and magical
```

**Reply by eroth** [reply]
Ah, thanks so much! I was actually one whitespace away from figuring this out based on reading through the docs and past discussions—I had:
```bash
dasel -r json 'all().mapOf(author, author.login)' -w csv < ~/Desktop/file.json
```
Note the whitespace in between `author, author.login`. Thanks again for your help, much appreciated!

**Reply by TomWright** [reply]
Yeah the selectors are whitespace sensitive right now.

It is a planned improvement but haven't gotten round to it yet. No problem!

**Reply by chmac** [reply]
After an hour or so of searching, I finally found this post. Thanks!

I still don't really understand how `mapOf` works though. I don't get why I need each item listed twice!

What about adding a simple example like this to the docs for `mapOf`? I had the same use case as @eroth. I wanted to get an array of JSON documents, pick a few fields, and generate a CSV. This example would have helped me out massively if it were easily found in the docs.

Thanks for the work on dasel btw, it's a great tool.

**Reply by TomWright** [reply]
Hey @chmac, you can find the `mapOf` docs here: https://daseldocs.tomwright.me/functions/mapof

In short though, the arguments are in pairs of key and value.

In the example above the map key is `author` and the value will be whatever is resolved from the `author.login` selector.

Good shout RE the docs - I'll see if I can get the `mapOf` docs improved today

**Reply by TomWright** [reply]
I've updated the docs: https://daseldocs.tomwright.me/functions/mapof
Please let me know if you think there could be any other improvements

**Reply by chmac** [reply]
@TomWright The updated docs look awesome, I'm impressed by the near instant response! :-)

In the example about outputting to CSV, I'd suggest outputting both name and phone number, so that it's clearer to people how to get 2 fields in the CSV. Now that I see your examples I realise that the first of the 2 pairs in `mapOf` is a string and doesn't need to exist in the data, that makes sense!

**Reply by chmac** [reply]
What I had in mind was replacing this:

    dasel -r json -w csv 'all().mapOf(phone,phone).merge()'

With something like this:

    dasel -r json -w csv 'all().mapOf(firstName,name.first,phone,phone).merge()'

Then hopefully the output would be:

```
firstName,phone
Tom,07
Joe,07
```

**Reply by TomWright** [reply]
Good idea @chmac , I've just made that change
